### PR TITLE
docs: addition of param and result sections

### PR DESCRIPTION
**Pull request by hachreak**
* Improves documentation. (addresses #48)

Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>

### CODE REVIEW

File: invenio_userprofiles/forms.py [hunk 1]
```diff
@@ -42,7 +42,11 @@
 
 
 def strip_filter(text):
-    """Filter for trimming whitespace."""
+    """Filter for trimming whitespace.
+
+    :param text: The text to strip.
+    :returns: The text stripped.
```
Comment (alizeepace) [inline invenio_userprofiles/forms.py]: The stripped text

File: invenio_userprofiles/models.py [hunk 2]
```diff
@@ -80,21 +80,36 @@ def username(self):
 
     @username.setter
     def username(self, username):
-        """Set username."""
+        """Set username.
+
+        .. note:: The username will be converted in lowecase. The display name
+        will contains the original version.
```
Comment (alizeepace) [inline invenio_userprofiles/models.py]: converted to lowercase (preposition and typo to correct), will contain

File: invenio_userprofiles/models.py [hunk 3]
```diff
@@ -80,21 +80,36 @@ def username(self):
...
         ).one()
 
     @classmethod
     def get_by_userid(cls, user_id):
-        """Get profile by username."""
+        """Get profile by username.
+
+        :param user_id: The :class:`invenio_accounts.models.User` ID.
+        :returns: A :class:`invenio_userprofiles.models.UserProfile` instance
+            on ``None``.
```
Comment (alizeepace) [inline invenio_userprofiles/models.py]: or ``None``

File: docs/conf.py [hunk 4]
```diff
@@ -328,4 +328,6 @@
 
 
 # Example configuration for intersphinx: refer to the Python standard library.
-intersphinx_mapping = {'https://docs.python.org/': None}
+intersphinx_mapping = {
+    'python-3.4': ('https://docs.python.org/3', None)
```
Comment (jirikuncar) [inline docs/conf.py]: @hachreak can you use 2.7, 3.5 or "latest"?

File: invenio_userprofiles/api.py [hunk 5]
```diff
@@ -34,6 +34,14 @@
 
 
 def _get_current_userprofile():
+    """Get current user profile.
+
+    .. note:: If the user is anonymous, then a
+    :class:`invenio_userprofiles.models.AnonymousUserProfile` instance is
```
Comment (jirikuncar) [inline invenio_userprofiles/api.py]: can you verify the indentation here?

File: invenio_userprofiles/models.py [hunk 6]
```diff
@@ -80,21 +80,38 @@ def username(self):
 
     @username.setter
     def username(self, username):
-        """Set username."""
+        """Set username.
+
+        .. note:: The username will be converted to lowecase. The display name
+        will contain the original version.
```
Comment (jirikuncar) [inline invenio_userprofiles/models.py]: Can you check the indentation here too?

File: invenio_userprofiles/models.py [hunk 7]
```diff
@@ -80,21 +80,38 @@ def username(self):
 
     @username.setter
     def username(self, username):
-        """Set username."""
+        """Set username.
+
+        .. note:: The username will be converted to lowecase. The display name
```
Comment (alizeepace) [inline invenio_userprofiles/models.py]: lowercase

**Comment by hachreak** [comment]
ping @jirikuncar

File: invenio_userprofiles/ext.py [hunk 8]
```diff
@@ -49,18 +50,9 @@ def init_app(self, app):
...
-            'USERPROFILES_PROFILE_URL',
-            '/account/settings/profile/')
-
-        app.config.setdefault('USERPROFILES_EMAIL_ENABLED', True)
-
-        app.config.setdefault(
-            'USERPROFILES_PROFILE_TEMPLATE',
-            'invenio_userprofiles/settings/profile.html')
+        for k in dir(config):
+            if k.startswith('USERPROFILES_'):
```
Comment (jirikuncar) [inline invenio_userprofiles/ext.py]: It will not load `USERPROFILES = True` from `config.py`.

Comment (jirikuncar) [inline invenio_userprofiles/ext.py]: ` ... or k == 'USERPROFILES':`

Comment (hachreak) [inline invenio_userprofiles/ext.py]: ok.
There is a special reason why is not followed the convention `MODULE_XXX` for `'USERPROFILES'`?

Comment (jirikuncar) [inline invenio_userprofiles/ext.py]: Do you have a suggestion for a better name?

Comment (jirikuncar) [inline invenio_userprofiles/ext.py]: Is it even used anywhere?

Comment (hachreak) [inline invenio_userprofiles/ext.py]: I found that it used only in `invenio_theme/templates/invenio_theme/header_login.html`.
There is also `config['ACCOUNTS']` there.
Should I follow the example of `invenio-accounts` and leave [app.config.setdefault('ACCOUNTS', True)](https://github.com/inveniosoftware/invenio-accounts/blob/master/invenio_accounts/ext.py#L156) inside the `ext.py`?

Comment (jirikuncar) [inline invenio_userprofiles/ext.py]: Yes. Just leave the `setdefault` version.

File: docs/conf.py [hunk 9]
```diff
@@ -328,4 +328,6 @@
 
 
 # Example configuration for intersphinx: refer to the Python standard library.
-intersphinx_mapping = {'https://docs.python.org/': None}
+intersphinx_mapping = {
+    'python': ('https://docs.python.org/3.5', None)
```
Comment (jirikuncar) [inline docs/conf.py]: please use `'https://docs.python.org/'`

**Comment by jirikuncar** [comment]
@hachreak LGTM, please check the intersphinx url, and rebase. Thanks!

File: docs/conf.py (hunk 9 above)
Comment (hachreak) [inline docs/conf.py]: it looks like at least `https://docs.python.org/3` is needed.
See: https://travis-ci.org/hachreak/invenio-userprofiles/jobs/159818887

File: invenio_userprofiles/models.py [hunk 10]
```diff
@@ -80,21 +80,38 @@ def username(self):
...
+        :returns: A :class:`invenio_userprofiles.models.UserProfile` instance.
+        """
         return cls.query.filter(
             UserProfile._username == username.lower()
         ).one()
 
     @classmethod
     def get_by_userid(cls, user_id):
-        """Get profile by username."""
+        """Get profile by username.
```
Comment (jirikuncar) [inline invenio_userprofiles/models.py]: by user identifier.

File: invenio_userprofiles/models.py [hunk 11]
```diff
@@ -80,21 +80,38 @@ def username(self):
 
     @username.setter
     def username(self, username):
-        """Set username."""
+        """Set username.
+
+        .. note:: The username will be converted to lowercase. The display name
+            will contain the original version.
+
+        :param username: The user name.
```
Comment (jirikuncar) [inline invenio_userprofiles/models.py]: please remove the `:param`

File: invenio_userprofiles/models.py [hunk 12]
```diff
@@ -80,21 +80,38 @@ def username(self):
...
         self._displayname = username
 
     @classmethod
     def get_by_username(cls, username):
-        """Get profile by username."""
+        """Get profile by username.
+
+        .. note:: The username is not case sensitive.
+
+        :param username: The user name.
```
Comment (jirikuncar) [inline invenio_userprofiles/models.py]: https://github.com/inveniosoftware/invenio-userprofiles/pull/54/files#r78776898

File: invenio_userprofiles/models.py [hunk 13]
```diff
@@ -80,21 +80,38 @@ def username(self):
...
+        :returns: A :class:`invenio_userprofiles.models.UserProfile` instance.
```
Comment (jirikuncar) [inline invenio_userprofiles/models.py]: Please remove it too ...

File: invenio_userprofiles/validators.py [hunk 14]
```diff
@@ -41,8 +42,11 @@
 def validate_username(username):
     """Validate the username.
 
-    Usernames can only contain numbers and latin letters. They must
-    start with a letter.
+    .. note:: Usernames can only contain numbers and latin letters and they
```
Comment (jirikuncar) [inline invenio_userprofiles/validators.py]: It would be useful to reference `username_regex`.

File: invenio_userprofiles/validators.py [hunk 15]
```diff
@@ -41,8 +42,11 @@
...
+        must start with a letter.
+
+    :param username: The user name.
```
Comment (jirikuncar) [inline invenio_userprofiles/validators.py]: no needed

File: invenio_userprofiles/validators.py [hunk 16]
```diff
@@ -41,8 +42,11 @@
...
+    :raises ValueError: If validation fails.
```
Comment (jirikuncar) [inline invenio_userprofiles/validators.py]: It's clear from the 2 lines of code

File: docs/conf.py [hunk 17]
```diff
@@ -326,9 +326,10 @@
 # If true, do not generate a @detailmenu in the "Top" node's menu.
 #texinfo_no_detailmenu = False
 
-
 # Example configuration for intersphinx: refer to the Python standard library.
-intersphinx_mapping = {'https://docs.python.org/': None}
+intersphinx_mapping = {
+    'python': ('https://docs.python.org/3', None)
```
Comment (jirikuncar) [inline docs/conf.py]: `'https://docs.python.org/'`

Comment (hachreak) [inline docs/conf.py]: as I wrote before, it looks like at least https://docs.python.org/3 is needed.
See: https://travis-ci.org/hachreak/invenio-userprofiles/jobs/159818887
Any solution?

Comment (jirikuncar) [inline docs/conf.py]: The problem seems to be caused by `:raises ...AssetionError:` which is not correct anyway and `assert` statement should be changed to `<module>Error`, `RuntimeError` or `ValueError`.

Comment (hachreak) [inline docs/conf.py]: Actually the problem looks like not the `AssertionError` but the `ValueError`.
`WARNING: py:exc reference target not found: ValueError`

File: invenio_userprofiles/validators.py (hunk 16 above)
Comment (hachreak) [inline invenio_userprofiles/validators.py]: It's true, but from documentation page people don't know which exception is.

**Pull request merged**